### PR TITLE
[12.0] easy my coop bond

### PR DESCRIPTION
**Pull request by houssine78**
This PR bring most of the data model for bond issue management. Is a first version some functionalities may be added.

### CODE REVIEW

File: easy_my_coop_loan/models/loan.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        self.write({'state': 'closed'})
+
+    @api.multi
+    def compute_loan_interest(self):
+        self.ensure_one()
+        interest_obj = self.env['loan.interest.line']
+        if self.interest_payment == 'end':
+            due_date = self.term_date
+        for line in self.loan_issue_lines:
+            # TODO remove this line
```
Comment (robinkeunen) [inline easy_my_coop_loan/models/loan.py:173]: pas oublier

File: easy_my_coop_loan/models/loan.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                accrued_taxes += taxes_amount
+                vals['interest'] = interest
+                vals['net_interest'] = net_interest
+                vals['taxes_amount'] = taxes_amount
+                vals['accrued_amount'] = accrued_amount
+                vals['accrued_interest'] = accrued_interest
+                vals['accrued_net_interest'] = accrued_net_interest
+                vals['accrued_taxes'] = accrued_taxes
+                vals['name'] = year
+                print(str(vals) + ' / ' + str(year))
```
Comment (robinkeunen) [inline easy_my_coop_loan/models/loan.py]: retirer le print.

File: easy_my_coop_loan_website/controllers/main.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
+from odoo import http
+from odoo.http import request
+
+from odoo.tools.translate import _
+
+
+class WebsiteLoanIssueSubscription(http.Controller):
+
+    @http.route(['/subscription/get_loan_issue'],
```
Comment (robinkeunen) [inline easy_my_coop_loan_website/controllers/main.py:9]: `subscription/loan_issues` would be a closer naming to classic api paths.

Tu peux pas faire un get pour un call ajax?

File: easy_my_coop_loan_website/controllers/main.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            return {
+                loan_issue.id: {
+                    'max_amount': loan_issue.maximum_amount_per_sub,
+                    'face_value': loan_issue.face_value,
+                    }
+                }
+        else:
+            return False
+
+    @http.route(['/page/loanissuesubscriptionform',
```
Comment (robinkeunen) [inline easy_my_coop_loan_website/controllers/main.py]: it'd be great to have a consistent naming with Rémy's pages. Such as

`/struct/<int:struct_id>/loan/subscription`

File: easy_my_coop_loan_website/controllers/main.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            return request.render(redirect, values)
+        if sub_amount:
+            values["error_msg"] = _("The amount shoud be of monetary type")
+            return request.render(redirect, values)
+        return True
+
+    @http.route(['/subscription/subscribe_loan_issue'],
+                type='http',
+                auth="user", website=True)
+    def loan_issue_subscription(self, **kwargs):
```
Comment (robinkeunen) [inline easy_my_coop_loan_website/controllers/main.py:78]: For consistency, it'd be great to use the same path as the form path. Then use the request type to diffentiate (GET / POST).

```python
    @http.route(['/struct/<int:struct_id>/loan/subscription'],
                type='http',
                auth="user",
                methos="GET",
                website=True)

    @http.route(['/struct/<int:struct_id>/loan/subscription'],
                type='http',
                auth="user",
                methos="POST",
                website=True)
```

**Review by robinkeunen — CHANGES_REQUESTED** [review]
Yo,

C'est très difficile de faire une review pertinente sur une si grosse PR. Du coup les commentaires sont très _cosmétiques_ et superficiels... N'hésite pas à nous soumettre des PR sur des commits fonctionnels dans le futur, nos reviews seront plus pertinentes.

Du coup, mon commentaire sur les noms des routes est le seul vraiment pertinent.

Je vais te faire des PR sur cette branche-ci avec des tests.

File: easy_my_coop_loan_website/controllers/main.py (hunk 4 above)
Comment (houssine78) [reply easy_my_coop_loan_website/controllers/main.py]: @robinkeunen this is the to be used in stand alone not in the platform so I'm not sure about the the asked form of the route.

File: easy_my_coop_loan_website/controllers/main.py (hunk 3 above)
Comment (houssine78) [reply easy_my_coop_loan_website/controllers/main.py:9]: except that this route return one loan_issue based on its id.

this is a json call so what do you mean by ajax ?

**Comment by robinkeunen** [comment]
Before merging, rebase on 12.0 and

- remove my commits 7a76a52 and 6ebbc5d, sorry, I pushed data files on 12.0
- squash your commits
- squash your commits to shorten history (commit messages are not all relevant here)

File: easy_my_coop_loan_website/controllers/main.py (hunk 4 above)
Comment (robinkeunen) [reply easy_my_coop_loan_website/controllers/main.py]: Well I think it'd be great to have consistent route naming regardless of the use.

File: easy_my_coop_loan_website/controllers/main.py (hunk 3 above)
Comment (robinkeunen) [reply easy_my_coop_loan_website/controllers/main.py:9]: You use this route in an ajax call (js calling the backend). I have a vague souvenir that you must use post methods in those calls. But if not, a get with url parameter would be more appropriate. `GET` to get data, `POST` to post data.

File: easy_my_coop_loan_website/controllers/main.py (hunk 5 above)
Comment (remytms) [reply easy_my_coop_loan_website/controllers/main.py:78]: This routes will go into `iwp_website` not in `easy_my_coop_loan_website`. :)

**Comment by remytms** [comment]
Would be great if this PR can be merge before the scrum of tomorrow. :)

File: easy_my_coop_loan_website/controllers/main.py (hunk 5 above)
Comment (robinkeunen) [reply easy_my_coop_loan_website/controllers/main.py:78]: @remytms So what naming would be good for you?

File: easy_my_coop_loan/models/loan.py (hunk 2 above)
Comment (houssine78) [reply easy_my_coop_loan/models/loan.py]: done

File: easy_my_coop_loan/models/loan.py (hunk 1 above)
Comment (houssine78) [reply easy_my_coop_loan/models/loan.py:173]: after the test. I still need it for now

File: easy_my_coop_loan_website/controllers/main.py (hunk 3 above)
Comment (houssine78) [reply easy_my_coop_loan_website/controllers/main.py:9]: Ok I changed it to GET

File: easy_my_coop_loan_website/controllers/main.py (hunk 5 above)
Comment (houssine78) [reply easy_my_coop_loan_website/controllers/main.py:78]: ok i changed the routes now

File: easy_my_coop_loan_website/controllers/main.py (hunk 4 above)
Comment (houssine78) [reply easy_my_coop_loan_website/controllers/main.py]: done

File: easy_my_coop_loan_website/controllers/main.py (hunk 3 above)
Comment (robinkeunen) [reply easy_my_coop_loan_website/controllers/main.py:9]: Great, the js is still working?

Comment (houssine78) [reply easy_my_coop_loan_website/controllers/main.py:9]: no

**Review by robinkeunen — APPROVED** [review]

**Pull request merged**